### PR TITLE
Adding more pipelines - E2E - Deploy shared env

### PR DESCRIPTION
**Pull request by julienstroheker**
## deploy-dev-env.yml
Used to deploy shared environment on demands
## pr-e2e.yml
Used to build/run the RP - Deploy a cluster against the RP - Run e2e

### CODE REVIEW

File: .pipelines/deploy-dev-env.yml [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        parameters:
+          gobin: ${{ variables.GOBIN }}
+          gopath: ${{ variables.GOPATH }}
+          goroot: ${{ variables.GOROOT }}
+          modulePath: ${{ variables.modulePath }}
+      - template: ./templates/template-setup-azure-tools.yml
+        parameters:
+          workingDirectory: ${{ variables.modulePath }}
+      - script: |
+          echo $(aro-v4-e2e-devops-spn) | base64 -d > aro-v4-e2e-devops-spn.json
```
Comment (mjudeikis) [inline .pipelines/deploy-dev-env.yml]: Is these visible in stdout? do we need set +x , -x wrappers?

Comment (julienstroheker) [reply .pipelines/deploy-dev-env.yml]: Secrets will be printed as `*****` in the console. 
We (MSFT + RH) are the only one having access to CI logs.

Was also planning to re add some `>/dev/null`

File: .pipelines/templates/template-run-rp.yml [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        --master-subnet "$CLUSTER-master" \
+        --worker-subnet "$CLUSTER-worker" \
+        --client-id $(cat devops-spn.json | jq -r '.clientId') \
+        --client-secret $(cat devops-spn.json | jq -r '.clientSecret')
+
+      echo "########## CLI : ARO List ##########"
+      az aro list -o table
+      echo "########## CLI : ARO list-creds ##########"
+      az aro list-credentials -g "$RESOURCEGROUP" -n "$CLUSTER"
+      echo "########## Run E2E ##########"
```
Comment (asalkeld) [inline .pipelines/templates/template-run-rp.yml]: just before this we need 
https://github.com/Azure/ARO-RP/blob/master/hack/e2e.sh#L75
and export KUBECONFIG

File: .pipelines/templates/template-run-rp.yml [hunk 3]
```diff
@@ -0,0 +1,83 @@
+parameters:
+  workingDirectory: ""
+steps:
+  - script: |
```
Comment (asalkeld) [inline .pipelines/templates/template-run-rp.yml]: is "set -xe" implied? what happens is one command fails? does it just keep running?

File: .pipelines/templates/template-run-rp.yml [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+      cd ${{ parameters.workingDirectory }}
+
+      . secrets/env
+      echo "########## Building ARO RP Binary ##########"
+      go build ./cmd/aro
```
Comment (asalkeld) [inline .pipelines/templates/template-run-rp.yml]: I assume this is on a clean vm? and it doesn't matter if this process is running after the test has completed?

**Comment by jim-minter** [comment]
@julienstroheker ping me when you're ready for a merge on this

**Comment by jim-minter** [comment]
another thing we do on the openshift side is either mark as draft (doesn't work for writing CI-related PRs) or adding "[WIP]" to the PR subject to avoid an accidental early merge

File: .pipelines/templates/template-run-rp.yml (hunk 4 above)
Comment (julienstroheker) [reply .pipelines/templates/template-run-rp.yml]: indeed

File: .pipelines/templates/template-run-rp.yml (hunk 3 above)
Comment (julienstroheker) [reply .pipelines/templates/template-run-rp.yml]: yep keep running for now. We can improve this for sure. I just want to make sure we are reaching out `az group delete` command at the end to clean as much as possible.

**Comment by julienstroheker** [comment]
@jim-minter after #122 will be merged, are we expecting CI creating a new SPN all the time for each cluster creation ?

**Comment by jim-minter** [comment]
> @jim-minter after #122 will be merged, are we expecting CI creating a new SPN all the time for each cluster creation ?

Discussed on a call - ideally yes, but if it's not feasible short term than finding a neat way of dropping the requirement (e.g. perhaps deploying a DB without the relevant unique key constraint) is acceptable.

**Comment by jim-minter** [comment]
@julienstroheker, there are some big shell scripts inside yaml files here and I'm worried about future maintainability and about duplication (e.g. with the instructions in deploy-development-rp.md).

Is it possible to pursue a path where for the bigger items (at least template-deploy-shared-env.yml, 
template-run-rp-and-e2e.yml, maybe template-setup-azure-tools.yml and template-prepare-secrets-env.yml as well), we can have these as free-standing shell scripts in hack/ and reference those scripts both from the CI yamls and from the markdown documents?

In v3 we found it invaluable for engineers to be able to run scripts like this directly from the laptop as part of the debugging process, not to have to rely on rerunning the pipeline all the time.

WDYT?

**Comment by julienstroheker** [comment]
Yea totally agreed on this. Was doing this to tests.
As soon as I validated the whole thing running and working, I’ll do the move. This should be quick.

**Comment by julienstroheker** [comment]
Also FYI on the status of this, Current hosted agent does not allow operation running for more than 1 hour. So I’ll need to use custom agent to validate this.

File: .pipelines/templates/template-run-rp-and-e2e.yml [hunk 5]
```diff
@@ -0,0 +1,81 @@
+parameters:
+  workingDirectory: ""
+steps:
+  - script: |
+      cd ${{ parameters.workingDirectory }}
+
```
Comment (asalkeld) [inline .pipelines/templates/template-run-rp-and-e2e.yml]: is it not worth merging this with hack/e2e.sh so that we can also run the test easily locally?

**Comment by julienstroheker** [comment]
@asalkeld @jim-minter can i have a review please ?

**Comment by julienstroheker** [comment]
You can see how this looks like here : https://dev.azure.com/msazure/AzureRedHatOpenShift/_build/results?buildId=28892764&view=logs&j=10f66467-59e5-53bc-0c8a-42cc89770de4&t=9dc613f3-d0be-54ed-3364-c030a6f93219

**Comment by mjudeikis** [comment]
@jim-minter @julienstroheker can we merge this as its already in un-reviewable state :/ 
I somehow think we should "merge, fix later" for this one :/

File: hack/e2e/run-rp-and-e2e.sh [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+      --vnet dev-vnet \
+      --master-subnet "$CLUSTER-master" \
+      --worker-subnet "$CLUSTER-worker"
+
+    echo "########## CLI : ARO List ##########"
+    az aro list -o table
+    echo "########## CLI : ARO list-creds ##########"
+    az aro list-credentials -g "$RESOURCEGROUP" -n "$CLUSTER"
+    echo "########## Run E2E ##########"
+    go run ./hack/kubeadminkubeconfig "/subscriptions/$AZURE_SUBSCRIPTION_ID/resourceGroups/$RESOURCEGROUP/providers/Microsoft.RedHatOpenShift/openShiftClusters/$CLUSTER" > ~/.kube/config
```
Comment (asalkeld) [inline hack/e2e/run-rp-and-e2e.sh]: my only minor quibble is to define and use KUBECONFIG rather than write over .kubeconfig as we use k8s/openshift for other things and over writing it is going to get annoying. anyways, can be a separate PR

**Comment by jim-minter** [comment]
@julienstroheker and I caught up by phone.  This is a great start.  There are a few areas with room for improvement, let's do some in this PR and open backlog items for the others.  General direction points:

* standardise secrets/ management across CI, RH and MSFT.  Use KV to store the contents of the secrets/ directory including env file.
* reduce duplication between CI templates / shell scripts / documentation.
* try to pull out things like template-deploy-shared-env.yml more into a shell script
* (probably a follow-up) get to a standard reference Dockerfile/container for running CI - standardises and avoids apt-gets, etc.  Might be nice for that to be RH-distro based too if possible (we can help with that) ;-)

File: hack/e2e/run-rp-and-e2e.sh [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+      --template-file deploy/databases-development.json \
+      --parameters \
+        "databaseAccountName=$COSMOSDB_ACCOUNT" \
+        "databaseName=$DATABASE_NAME" \
+        >/dev/null
+}
+
+deploy_e2e_deps() {
+    mkdir -p ~/.kube
+    RESOURCEGROUP="v4-e2e-rg-$(git log --format=%h -n 1 HEAD)-$LOCATION"
```
Comment (JackQuincy) [inline hack/e2e/run-rp-and-e2e.sh]: these are duplicated can we share setting these variables?

**Comment by julienstroheker** [comment]
@jim-minter i did a new iteration based on what we talked.
Let me know if i missed something. 
Thanks

File: Makefile [hunk 8]
```diff
@@ -70,12 +70,18 @@ pyenv${PYTHON_VERSION}:
 		sed -i -e "s|^dev_sources = $(PWD)$$|dev_sources = $(PWD)/python|" ~/.azure/config
 
 secrets:
+	@[ "${SECRET_SA_ACCOUNT_NAME}" ] || ( echo ">> SECRET_SA_ACCOUNT_NAME is not set"; exit 1 )
```
Comment (julienstroheker) [inline Makefile]: Using a storage account + AAD auth works like a charm.

**Comment by jim-minter** [comment]
:tada:
